test/jdk/java/nio/channels/DatagramChannel/Connect.java

```diff
@@ -43,89 +47,92 @@ public static void main(String[] args) throws Exception {
     }
 
     static void test() throws Exception {
-        Reactor r = new Reactor();
-        Actor a = new Actor(r.port());
-        invoke(a, r);
+        try (Reactor r = new Reactor();
+             Actor a = new Actor(r.port())) {
+            invoke(a, r);
+        }
     }
 
-    static void invoke(Sprintable reader, Sprintable writer) throws Exception {
-
-        Thread writerThread = new Thread(writer);
-        writerThread.start();
-
-        Thread readerThread = new Thread(reader);
-        readerThread.start();
-
-        writerThread.join();
-        readerThread.join();
-
-        reader.throwException();
-        writer.throwException();
+    static void invoke(Runnable reader, Runnable writer) throws CompletionException {
+        ExecutorService threadPool = Executors.newCachedThreadPool();
+        try {
+            CompletableFuture<Void> f1 = CompletableFuture.runAsync(writer, threadPool);
+            CompletableFuture<Void> f2 = CompletableFuture.runAsync(reader, threadPool);
+            wait(f1, f2);
+        } finally {
+            threadPool.shutdown();
+        }
```

Comment:
I believe it would be better to call threadPool.shutdown() **after** having closed the Actor/Reactor: closing the Actor/Reactor makes sure that both thread terminates - and shutdown() will wait for the threads to terminate. Therefore I'd suggest to create the threadPool in the `test` method instead, and pass it as first parameter to the `invoke` method.

Reply:
As you suggested, I moved the threadPool.shutdown() call to the main test method instead of having it in invoke. While it seems that shutdown() does conduct an orderly shutdown of the pool, I agree that its best to be safe and to make absolutely certain that Actor/Reactor close properly..